bot/bot.go

```diff
@@ -32,42 +32,95 @@ type Bot interface {
-func NewBot(host string, ssl bool, nick, ident string, channels []string) Bot {
-	return NewBotWithProxy(host, ssl, nick, ident, channels, "")
-}
-
-// NewBotWithProxy creates a new Bot implementation with a set of parameters and a proxy.
-func NewBotWithProxy(host string, ssl bool, nick, ident string, channels []string, proxy string) Bot {
+// NewBotWithProxyPassword creates a new Bot implementation with required parameters and options.
+func NewBotOptions(host string, nick string, options ...func(*BotImpl)) (Bot, error) {
 	hostPort := strings.SplitN(host, ":", 2)
 	cfg := &client.Config{
-		Me:          &state.Nick{Nick: nick},
+		Me: &state.Nick{
+			Nick:  nick,
+			Ident: nick,
+			Name:  nick,
+		},
 		NewNick:     func(s string) string { return s + "_" },
 		PingFreq:    3 * time.Minute,
 		QuitMessage: "I have to go.",
 		Server:      host,
-		SSL:         ssl,
+		SSL:         false,
 		SSLConfig:   tls.Config(hostPort[0]),
 		Version:     "Powered by GoIRCBot",
 		Recover:     (*client.Conn).LogPanic,
 		SplitLen:    450,
-		Proxy:       proxy,
+		Proxy:       "",
+		Pass:        "",
 	}
-	cfg.Me.Ident = ident
-	cfg.Me.Name = nick
 
-	conn := client.Client(cfg)
-	conn.EnableStateTracking()
 	b := &BotImpl{
-		conn:      conn,
+		config:    cfg,
 		reconnect: true,
 		quit:      make(chan bool),
 		commands:  NewCommands(),
-		channels:  channels,
 	}
 
+	for _, option := range options {
+		option(b)
+	}
+
+	return b, nil
+}
+
+func SSL(ssl bool) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.SSL = ssl }
+}
+
+func Ident(ident string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Me.Ident = ident }
+}
+
+func RealName(realName string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Me.Name = realName }
+}
+
+func Proxy(proxy string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Proxy = proxy }
+}
+
+func Password(password string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Pass = password }
+}
+
+func Channels(channels []string) func(*BotImpl) {
+	return func(b *BotImpl) { b.channels = channels }
+}
+
+// NewBot creates a new Bot implementation with a set of parameters.
+func NewBot(host string, ssl bool, nick, ident string, channels []string) Bot {
+	b, _ := NewBotOptions(host, nick, SSL(ssl), Ident(ident), Channels(channels))
+	return b
+}
+
+// NewBotWithProxy creates a new Bot implementation with a set of parameters including a proxy.
+func NewBotWithProxy(host string, ssl bool, nick, ident string, channels []string, proxy string) Bot {
+	b, _ := NewBotOptions(host, nick, SSL(ssl), Ident(ident), Proxy(proxy), Channels(channels))
+	return b
+}
+
+// BotImpl implements Bot.
+type BotImpl struct {
+	config    *client.Config
+	conn      *client.Conn
+	reconnect bool
+	quit      chan bool
+	commands  *Commands
+	channels  []string
+}
+
+// Run starts the Bot by connecting it to IRC. It automatically reconnects.
+func (b *BotImpl) Run() {
+	b.conn = client.Client(b.config)
```

Comment:
Good choice of later initialization and set up of internal handlers in `Run()`.

I did like the function body of `Run()` being small and focused on just the for loop though, maybe this could be moved to an internal `setup()` that `Run()` calls at the beginning, what do you think?

Reply:
Fine with me. Just `b.conn = setup(b)`?

Reply:
Ah you mean doing everything except `for ... {}` in a `setup()` func

Reply:
Not `b.conn = setup(b)`, more generic: move the whole block that you moved there into a `func (* BotImpl) setup()` and call it from `Run()`.

Reply:
Like you did :)